### ISSUE TITLE
issue with "use_belief_predictor"

### ISSUE THREAD

**Issue by kiitaamuuraa** [facebookresearch/sound-spaces#51]
Hello, I appreciate your wonderful work.

I tried to rerun your ```savi``` experiments following the [usage](https://github.com/facebookresearch/sound-spaces/tree/master/ss_baselines/savi#usage) you wrote. However, an error occurred at l.203 of ```ppo_trainer.py``` when I tried to run ```python ss_baselines/savi/run.py --exp-config ss_baselines/savi/config/semantic_audionav/savi.yaml --model-dir data/models/savi```. 
https://github.com/facebookresearch/sound-spaces/blob/9e3318b5e54f3246a606d2373ca60e4b7efc11f1/ss_baselines/savi/ppo/ppo_trainer.py#L203
The error was ```KeyError: 'belief_predictor'```. In fact, I made sure that ```ckpt.399.pth```, which had been created in pre-training, did not have the key named ```belief_predictor```.  Attached is the [error.log](https://github.com/facebookresearch/sound-spaces/files/6962541/error.log).
**How should I pretrain ```savi``` whose ```chpt.XXX.pth``` stores ```belief_predictor```?**

Also, I changed ```use_belief_predictor``` in ```savi.yaml``` from ```True``` to ```False```. The training itself worked without error, but the result was quite worse than what you reported in your [paper](https://arxiv.org/abs/2012.11583).
https://github.com/facebookresearch/sound-spaces/blob/9e3318b5e54f3246a606d2373ca60e4b7efc11f1/ss_baselines/savi/config/semantic_audionav/savi.yaml#L33
**Is that a reasonable result for you?**

**Comment by ChanganVR** [comment]
Hi @kiitaamuuraa  the pretraining config was wrong and I have fixed them. With the pushed changes, it should work now.

**Comment by kiitaamuuraa** [comment]
Hello @ChanganVR 
Thanks for your modification. The pre-training worked well with your modified newer configuration files, I think. However, while I got reasonable results, concretely spl>=0.20, Success>=0.30, sna>=0.15, in the pre-training, the performances from the second stage training, which is conducted with the following command you mention in the README.md. significantly dropped. 
`
python ss_baselines/savi/run.py --exp-config ss_baselines/savi/config/semantic_audionav/savi.yaml --model-dir data/models/savi`

Are there any reasons you know that cause the lower performance?

Also, I saw the memory size is 150 from the pretraining configuration, `savi_pretraining.yaml`, and it makes me confused because you write size of the memory is 1 in pretraining in the research paper.

https://github.com/facebookresearch/sound-spaces/blob/9e3318b5e54f3246a606d2373ca60e4b7efc11f1/ss_baselines/savi/config/semantic_audionav/savi_pretraining.yaml#L40

**Comment by ChanganVR** [comment]
@kiitaamuuraa  you're right, the memory_size should be 1 for the pretraining. It's very likely I used the savi.yaml for savi_pretraning.yaml and that's why there were so many bugs to fix. Sorry about that :( 

>the performances from the second stage training , which is conducted with the following command you mention in the README.md. significantly dropped.    

Do you mean the performance in this stage dropped compared to the pretrained model? And may I know how many steps did you train the savi model and how many GPUs did you use (batch size) in the second stage?

**Comment by kiitaamuuraa** [comment]
Thank you for the quick response. For the first topic, is it better for me to set  `memory_size: 1` in `savi_pretrain.yaml` if I want to reproduce savi completely?


For the second topic, I do; you are right. I used one GPU. I am not sure how many steps I trained in the second stage; according to the `train.log`, 30 steps? Anyway, I did not modify any lines  except for  `pretrained_weights: "data/models/savi/data/ckpt.XXX.pth"` in `savi.yaml`.

Here is the train.log of the second stage. I hope it helps you find anything wrong.

> 2021-08-16 12:31:30,720 update: 19960	fps: 195.849	
> 2021-08-16 12:31:30,720 update: 19960	env-time: 20.729s	pth-time: 33.768s	frames: 23953200.0
> 2021-08-16 12:31:30,720 Average window size: 10  distance_to_goal: 9.889  na: 201.685  normalized_distance_to_goal: 0.616  reward: 2.150  sna: 0.022  softspl: 0.263  spl: 0.072  success: 0.111  sws: 0.093
> 2021-08-16 12:32:29,554 update: 19970	fps: 199.823	
> 2021-08-16 12:32:29,554 update: 19970	env-time: 38.768s	pth-time: 67.793s	frames: 23965200.0
> 2021-08-16 12:32:29,554 Average window size: 20  distance_to_goal: 9.940  na: 152.735  normalized_distance_to_goal: 0.591  reward: 3.022  sna: 0.036  softspl: 0.275  spl: 0.086  success: 0.119  sws: 0.099
> 2021-08-16 12:33:37,327 update: 19980	fps: 191.612	
> 2021-08-16 12:33:37,328 update: 19980	env-time: 65.364s	pth-time: 102.180s	frames: 23977200.0
> 2021-08-16 12:33:37,328 Average window size: 30  distance_to_goal: 10.122  na: 124.824  normalized_distance_to_goal: 0.571  reward: 3.680  sna: 0.050  softspl: 0.319  spl: 0.116  success: 0.144  sws: 0.115
> 2021-08-16 12:34:37,438 update: 19990	fps: 193.556	
> 2021-08-16 12:34:37,438 update: 19990	env-time: 84.150s	pth-time: 136.711s	frames: 23989200.0
> 2021-08-16 12:34:37,438 Average window size: 40  distance_to_goal: 10.810  na: 105.545  normalized_distance_to_goal: 0.591  reward: 3.766  sna: 0.062  softspl: 0.316  spl: 0.112  success: 0.139  sws: 0.102
> 2021-08-16 13:01:13,170 Overwriting CNN input size of depth: (64, 64)
> 2021-08-16 13:01:13,170 Overwriting CNN input size of rgb: (64, 64)
> 2021-08-16 13:01:15,263 agent number of trainable parameters: 1197413
> 2021-08-16 13:01:15,264 belief predictor number of trainable parameters: 710530
> 2021-08-16 13:01:15,264 config: BASE_TASK_CONFIG_PATH: configs/semantic_audionav/savi/mp3d/

**Comment by ChanganVR** [comment]
>For the first topic, is it better for me to set memory_size: 1 in savi_pretrain.yaml if I want to reproduce savi completely?

Yes, I have already pushed the changes.

>For the second topic, I do; you are right. I used one GPU. I am not sure how many steps I trained in the second stage;

Based on your log, it looks like you have trained for 23M steps so far and the performance is still increasing. We trained the model for 300M to fully converge as writen in the paper. Can you plot the curve with tensorboard? Does the training/validation curve look like converged to you?

**Comment by spaul007** [comment]
How much time does it take for this 300M steps training and how many GPUs are you using?

**Comment by ChanganVR** [comment]
To speed up the training time, I used 32 GPUs to train for around 18 hours. I didn't experiment much with the number of GPUs, but I think reducing the number of GPUs should be fine at the cost of longer training time. In addition, 300M was more than sufficient for every model to fully converge. The performance started to saturate around 100M steps.

So in another word, if you have 8 GPUs, you should see the model converge in about a day. If you have 4 GPUs, it would take about 2 days and etc. I'm sorry if this is very compute expensive, visual embodied learning with RL generally requires a lot of GPU time to train...

**Comment by kiitaamuuraa** [comment]
Do you mean that I need to change this NUM_UPDATES?
https://github.com/facebookresearch/sound-spaces/blob/0e87180459a5c9901bd1b17fe83405ebe57b9360/ss_baselines/savi/config/semantic_audionav/savi.yaml#L5

**Comment by ChanganVR** [comment]
The total number of training steps are proportional to the number of GPUs and the default is 32 as specified in the `slurm.sh` script. If you reduce the number of GPUs, you need to increase the number of `NUM_UPDATES`.

**Comment by kiitaamuuraa** [comment]
I see. Unfortunately, as I do not use slurm cluster, I am increasing the NUM_UPDATES. If I finish my training successfully, I will report it here; so please allow me to keep this issue open for a while.
